### PR TITLE
Use device loopback IP address to send SNMP query from neighboring ceos or vsonic

### PR DESCRIPTION
**Pull request by SuvarnaMeenakshi**
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 201911
- [ ] 202012
- [ ] 202205

#### What is the motivation for this PR?
https://github.com/sonic-net/sonic-buildimage/pull/15487 modifies snmpd in SONiC to listen on management and loopback ip by default instead of listening on any IP.
test_interop_protocol.py::test_snmp sends a query to the neighbor device using the link ip address.
After the above change to listen on management and loopback ip, the snmp query will fail.
Hence, modifying the test to send SNMP query to SONiC DUT from neighboring vsonic/eos using Loopback IP of SONiC DUT.

#### How did you do it?
Get loopback IPv4 address DUT and use that to query from neighbor device.
Added a generic helped function to send SNMP query to DUT from a neighbor.

### CODE REVIEW

**Comment by abdosi** [comment]
@SuvarnaMeenakshi can we remove this check now ?

`if duthost.is_multi_asic:
            pytest.skip("The test is for Single ASIC devices")`

I wanted to make sure it run on multi-asic platforms also. Since we are using loobpack ip i think it should work

**Comment by SuvarnaMeenakshi** [comment]
> > @SuvarnaMeenakshi can we remove this check now ?
> > `if duthost.is_multi_asic: pytest.skip("The test is for Single ASIC devices")`
> > I wanted to make sure it run on multi-asic platforms also. Since we are using loobpack ip i think it should work
> 
> Sure, will test this out on multi-asic and make the change. snmp should have worked even with link ip on multi-asic, not sure why the skip was added.

@abdosi this will require additional work as this test is sending snmp query from multi-asic dut to neighbor vsonic.
From multi-asic DUT, the loopback or link ip to neighbor vsonic/ceos is not reachable that is why it is skipped here, it is reachable from front-end asic namespace.
```
admin@vlab-08:~$ ping 100.1.0.1
PING 100.1.0.1 (100.1.0.1) 56(84) bytes of data.
^C
--- 100.1.0.1 ping statistics ---
2 packets transmitted, 0 received, 100% packet loss, time 1012ms

admin@vlab-08:~$ sudo ip netns exec asic0 ping 100.1.0.1
PING 100.1.0.1 (100.1.0.1) 56(84) bytes of data.
64 bytes from 100.1.0.1: icmp_seq=1 ttl=64 time=1.12 ms
64 bytes from 100.1.0.1: icmp_seq=2 ttl=64 time=1.11 ms

--- 100.1.0.1 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1001ms
rtt min/avg/max/mdev = 1.107/1.111/1.115/0.004 ms
```
I will raise a github issue to fix this and enable this test on multi-asic/
Can we keep this PR scope to ensure current test case works as is.

**Comment by SuvarnaMeenakshi** [comment]
> test is sending snmp query from multi-asic dut to neighbor vsonic. From multi-asic DUT, the loopback or link ip to neighbor vsonic/ceos is not reachable that is why it is skipped here, it is reachable from front-end asic namespace.
> 
> ```
> admin@vlab-08:~$ ping 100.1.0.1
> PING 100.1.0.1 (100.1.0.1) 56(84) bytes of data.
> ^C
> --- 100.1.0.1 ping statistics ---
> 2 packets transmitted, 0 received, 100% packet loss, time 1012ms
> 
> admin@vlab-08:~$ sudo ip netns exec asic0 ping 100.1.0.1
> PING 100.1.0.1 (100.1.0.1) 56(84) bytes of data.
> 64 bytes from 100.1.0.1: icmp_seq=1 ttl=64 time=1.12 ms
> 64 bytes from 100.1.0.1: icmp_seq=2 ttl=64 time=1.11 ms
> 
> --- 100.1.0.1 ping statistics ---
> 2 packets transmitted, 2 received, 0% packet loss, time 1001ms
> rtt min/avg/max/mdev = 1.107/1.111/1.115/0.004 ms
> ```

https://github.com/sonic-net/sonic-mgmt/issues/8817

File: tests/macsec/test_interop_protocol.py [hunk 1]
```diff
@@ -132,11 +133,33 @@ def test_snmp(self, duthost, ctrl_links, upstream_links, creds, wait_mka_establi
                     commands=['show snmp community | include name'])
                 community = re.search(r'Community name: (\S+)',
                                       result['stdout'][0]).groups()[0]
+                up_link = upstream_links[ctrl_port]
+                nbr_ip = up_link["local_ipv4_addr"]
             else:  # vsonic neighbour
                 community = creds['snmp_rocommunity']
+                # Use LoopbackIP to query from vsonic neighbor
```
Comment (qiluo-msft) [inline tests/macsec/test_interop_protocol.py:140]: In the code comment, could you be more specific about ipv4 or ipv6?

File: tests/macsec/test_interop_protocol.py [hunk 2]
```diff
@@ -132,11 +133,33 @@ def test_snmp(self, duthost, ctrl_links, upstream_links, creds, wait_mka_establi
...
+                result = nbr["host"].config_facts(host=nbr["host"].hostname,
+                                                  source="running")[
+                                                  "ansible_facts"].get(
+                                                  "LOOPBACK_INTERFACE",
+                                                  {}).get('Loopback0', {})
+                for ip in result:
+                    if isinstance(ipaddress.ip_address(ip.split('/')[0]),
```
Comment (qiluo-msft) [inline tests/macsec/test_interop_protocol.py:147]: No need to split, you can directly parse `ipaddress.ip_network`

File: tests/macsec/test_interop_protocol.py [hunk 3]
```diff
@@ -132,11 +133,33 @@ def test_snmp(self, duthost, ctrl_links, upstream_links, creds, wait_mka_establi
...
+                nbr_ip = up_link["local_ipv4_addr"]
             else:  # vsonic neighbour
                 community = creds['snmp_rocommunity']
+                # Use LoopbackIP to query from vsonic neighbor
+                result = nbr["host"].config_facts(host=nbr["host"].hostname,
+                                                  source="running")[
+                                                  "ansible_facts"].get(
+                                                  "LOOPBACK_INTERFACE",
+                                                  {}).get('Loopback0', {})
+                for ip in result:
```
Comment (qiluo-msft) [inline tests/macsec/test_interop_protocol.py:146]: >for [](http://example.com/codeflow?start=16&length=3)

You may use `for-else` to capture the situation that no ip satisfy the condition. In the `else` block, you can explicitly fail the test.

File: tests/macsec/test_interop_protocol.py (hunk 1 above)
Comment (SuvarnaMeenakshi) [reply tests/macsec/test_interop_protocol.py]: Fixed

File: tests/macsec/test_interop_protocol.py (hunk 2 above)
Comment (SuvarnaMeenakshi) [reply tests/macsec/test_interop_protocol.py]: Fixed

File: tests/macsec/test_interop_protocol.py (hunk 3 above)
Comment (SuvarnaMeenakshi) [reply tests/macsec/test_interop_protocol.py:146]: Fixed

**Comment by yejianquan** [comment]
LGTM, @Pterosaur , can you also confirm?

**Comment by Pterosaur** [comment]
lgtm

**Comment by SuvarnaMeenakshi** [comment]
@abdosi modified the macsec/test_interop_protocol.py test_snmp to send SNMP query from neighboring vsonic/ceos devices to SONiC DUT. This way we can have the test running on multi-asic device as well.

While testing out on multi-asic device, found that there might be additional changes required. 
So will keep the skip for multi-asic as is in this PR and raise another PR with changes required for multi-asic.

**Review by abdosi — APPROVED** [review]

**Review by qiluo-msft — APPROVED** [review]

**Pull request merged**